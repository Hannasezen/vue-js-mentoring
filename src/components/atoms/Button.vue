<template>
  <button :class="classes" type="button">{{ name }}</button>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";

export default defineComponent({
  name: "Button",
  props: {
    name: {
      type: String,
      required: true,
    },
    buttonType: {
      type: String as PropType<string>,
      validator: function (value: string): boolean {
        return ["primary", "secondary", "tertiary"].includes(value);
      },
      default: "primary",
    },
  },
  computed: {
    classes(props) {
      return {
        button: true,
        [`button--${props.buttonType}`]: props.buttonType,
      };
    },
  },
});
</script>

<style lang="scss" scoped>
@import "@/assets/styles/main.scss";

.button {
  padding: 14px;
  border-radius: 4px;
  border: 2px solid $button-bg-color;
  font-family: $font-family;
  font-size: $font-size-m;
  line-height: $line-height-m;
  font-weight: 500;
  text-transform: uppercase;
  text-align: center;
  color: $font-color;
  background: $button-bg-color;
  cursor: pointer;
  transition: all ease 0.3s;

  &:hover {
    background: $button-hover-bg-color;
    border-color: $button-hover-bg-color;
  }

  &--secondary {
    color: $font-color-secondary;
    background: transparent;

    &:hover {
      color: $font-color;
    }
  }

  &--tertiary {
    padding: 11px;
    border: none;
    font-weight: 600;
    color: $font-color-secondary;
    background: rgba(96, 96, 96, 0.68);

    &:hover {
      color: $font-color;
    }
  }
}
</style>
