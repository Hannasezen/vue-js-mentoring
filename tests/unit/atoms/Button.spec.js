import { shallowMount } from "@vue/test-utils";
import Button from "@/components/atoms/Button.vue";

describe("Button.vue", () => {
  let defaultProps;
  let mockOnClickFunction;

  beforeEach(() => {
    mockOnClickFunction = jest.fn();

    defaultProps = {
      name: "Button",
      buttonType: "primary",
      type: "submit",
      onClick: mockOnClickFunction,
    };
  });

  it("renders properly", () => {
    const wrapper = shallowMount(Button, {
      props: defaultProps,
    });

    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  it("calls onClick() when button clicked", async () => {
    const wrapper = shallowMount(Button, {
      props: defaultProps,
    });

    await wrapper.trigger("click");

    expect(mockOnClickFunction).toBeCalledTimes(1);
  });
});
