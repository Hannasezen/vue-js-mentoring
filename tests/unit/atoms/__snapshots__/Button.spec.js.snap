// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue renders properly 1`] = `
<button
  class="button button--primary"
  type="submit"
>
  Button
</button>
`;
